Add scopes pages and refactor Expressions page (#947)
wso2/docs-mi: en/docs/reference/synapse-properties/expressions.md

# Expressions

In WSO2 Micro Integrator, expressions are a way to dynamically compute values within various integration artifacts such as sequences, APIs, endpoints, and proxies. These expressions are used to extract data as it flows through the integration logic.

The most commonly used expression languages in WSO2 MI are [Synapse Variables](#synapse-variables), [JSONPath (for JSON)](#jsonpath-expressions), and [XPath (for XML)](#xpath-expressions).

> ## Synapse Variables

Reviewer: ```suggestion
## Synapse variables
```

There is a set of predefined synapse variables that you can use
to extract data from the message. These synapse variables in different [scopes]({{base_path}}/reference/synapse-properties/scopes) are as follows:

### $ctx

> The prefix used to get properties set at the [default scope]({{base_path}}/reference/synapse-properties/scopes/#default-scope). For example, to get the value of property with name `ERROR_MESSAGE`, use the expression `$ctx:ERROR_MESSAGE`.

Reviewer: Shall we reword this as one of the following?

- For example, to get the value of the property with the `ERROR_MESSAGE` name, use the `$ctx:ERROR_MESSAGE` expression.

or

- For example, to get the value of the `ERROR_MESSAGE` property name, use the `$ctx:ERROR_MESSAGE` expression.

Author: No it won't express the fact that we highlight here.

> Similarly, you can use `$ctx` prefix with [Generic Properties]({{base_path}}/reference/mediators/property-reference/generic-properties).

Reviewer: ```suggestion
Similarly, you can use the `$ctx` prefix with [Generic Properties]({{base_path}}/reference/mediators/property-reference/generic-properties).
```

??? note "Example of $ctx usage"
    
    This example sends a request to an API, and call to a non-existent endpoint reference key. It causes a
    mediation fault, which triggers the fault sequence.

    Deploy the following API. For instructions, see [Creating a REST API]({{base_path}}/develop/creating-artifacts/creating-an-api/).
    

>     Note the property, `<property name="stockprop" expression="$ctx:ERROR_MESSAGE"/>` in the fault sequence configuration. It is used to log the error message that occurs due to a mediation fault.

Reviewer: ```suggestion
    The `<property name="stockprop" expression="$ctx:ERROR_MESSAGE"/>` property in the fault sequence configuration is used to log the error message that occurs due to a mediation fault.
```

                <call>
                    <endpoint key="ep2"/>
                </call>
                <respond/>   
            </inSequence>
            <faultSequence>
                <log>
                    <property name="stockprop" expression="$ctx:ERROR_MESSAGE"/>
                    <property name="Cause" expression="get-property('ERROR_MESSAGE')"/>
                </log>
            </faultSequence>
        </resource>
    </api> 
    ```
     
    Note the following message in the log.

    ``` java
    INFO {LogMediator} - {api:stockQuoteAPI} To: /stockQuote/, MessageID: urn:uuid:f1c35024-4d19-4042-abed-8880a4d98c3d, correlation_id: f1c35024-4d19-4042-abed-8880a4d98c3d, Direction: request, stockprop = Couldn't find the endpoint with the key : ep2, Cause = Couldn't find the endpoint with the key : ep2
    ```
    In this example, the property definition, `<property
    name="stockprop" expression="$ctx:ERROR_MESSAGE"/>` is equivalent
    to `<property name="stockprop"
    expression="get-property('ERROR_MESSAGE')"/>`.

### $trp

The prefix used to get the transport headers. For example, to get the
transport header named Content-Type of the current message, use the

> expression **$trp:Content-Type** . HTTP transport headers are not

Reviewer: ```suggestion
`$trp:Content-Type` expression . HTTP transport headers are not
```

    
    Note the property, `<property name="stockprop" expression="$trp:Content-Type"/>` in the configuration, which is used to log the `Content-Type` HTTP header of the request message.
    
    ``` xml
    <api context="/stockQuote" name="stockQuoteAPI" xmlns="http://ws.apache.org/ns/synapse">
        <resource methods="POST" uri-template="/">
            <inSequence>
                <log>
                    <property name="stockprop" expression="$trp:Content-Type"/>
                </log>
                <respond/>
            </inSequence>
            <faultSequence/>
        </resource>
    </api> 
    ```
    
    Note the following message in the log.
    
    ``` java
    INFO {LogMediator} - {api:stockQuoteAPI} To: /stockQuote/, MessageID: urn:uuid:7095f877-6a7c-4c99-a48b-c506ed9e43e1, correlation_id: 7095f877-6a7c-4c99-a48b-c506ed9e43e1, Direction: request, stockprop = application/json
    ```
    
    In this example, the property definition, `<property name="stockprop"
    expression="$trp:Content-Type"/>` is equivalent to `<property
    name="stockprop"
    expression="get-property('transport','Content-Type')"/>`.

### $body

> The prefix used to get the message payload. For example, the expression `$body//getQuote` refers to the first `getQuote` element in a message body.

Reviewer: ```suggestion
The prefix used to get the message payload. For example, the `$body//getQuote` expression refers to the first `getQuote` element in a message body.
```

                <log>
                    <property xmlns:m0="http://services.samples" name="stockprop" expression="$body//m0:getQuote"/>
                </log>
                <respond/>
            </inSequence>
        </target>
    </proxy> 
    ```
    
    Note the following message in the log.
    
    ``` java
    INFO - LogMediator To: /services/StockQuoteProxy, WSAction: urn:getQuote, SOAPAction: urn:getQuote, ReplyTo: http://www.w3.org/2005/08/addressing/anonymous, MessageID: urn:uuid:930f68f5-199a-4eff-90d2-ea679c2362ab, Direction: request, stockprop = <m0:getQuotexmlns:m0="http://services.samples"><m0:request><m0:symbol>IBM</m0:symbol></m0:request></m0:getQuote>
    ```

### $axis2

> The prefix used to get the properties at the [axis2 scope]({{base_path}}/reference/synapse-properties/scopes/#axis2-scope). For example, to get the value axis2 scope property with name `REST_URL_POSTFIX`, use the expression `$axis2:REST_URL_POSTFIX`.

Reviewer: ```suggestion
The prefix used to get the properties at the [axis2 scope]({{base_path}}/reference/synapse-properties/scopes/#axis2-scope). For example, to get the value of the `axis2` scope property with the `REST_URL_POSTFIX` name, use the `$axis2:REST_URL_POSTFIX` expression.
```

        ``` xml
        {"theData":{"item":{"content":{"paramA":"ValueA","paramB":"valueB"}}}}
        ```

## JSONPath expressions

[JSONPath](http://goessner.net/articles/JsonPath/) expressions are used for navigating and querying JSON data. JSONPath allows you to extract specific data from JSON documents.

JSONPath expressions are used within mediators like Log, Filter, Property, and Switch mediators to extract JSON data.
`json-eval` function in expressions is used to specify JSONPath.

> - The syntax usually starts with $ to represent the root of the JSON document.

Reviewer: ```suggestion
- The syntax usually starts with `$` to represent the root of the JSON document.
```

</tbody>
</table>

You can evaluate a JSONPath expression against a property that contains a JSON payload using [Synapse Variables](#synapse-variables).

To evaluate a JSONPath expression against a property, use the following syntax.

```json
json-eval(<scope_of_the_property>:<property_name>.<JSONPath_expression>)
```

Example 1: When the property is in the default scope you can use [$ctx](#ctx).

```json
json-eval($ctx:propertyName.student.name)
```

Example 2: When the property is in the Axis2 scope you can use [$axis2](#axis2).

```json
json-eval($axis2:propertyName.student.name)
```

Example 3: When the property is in the transport scope you can use [$trp](#trp).

```json
json-eval($trp:propertyName.student.name)
```

## XPath expressions

[XPath (XML Path Language)](https://www.w3schools.com/xml/xpath_intro.asp) is a language used to navigate through elements and attributes in an XML document. It allows you to identify and extract specific portions of XML data, making it ideal for working with complex XML payloads, which are common in service integration scenarios.

!!! Note

>     The WSO2 Micro Integrator supports standard XPath functions and variables through its underlying XPath engine. It supports XPath 1.0 by default where as the support for XPath 2.0 can be introduced by adding the following property in `<MI_HOME>/conf/deployment.toml`.

Reviewer: ```suggestion
    The WSO2 Micro Integrator supports standard XPath functions and variables through its underlying XPath engine. It supports XPath 1.0 by default whereas the support for XPath 2.0 can be introduced by adding the following property in `<MI_HOME>/conf/deployment.toml`.
```

    
    ```toml
    [mediation]
    synapse.enable_xpath_dom_failover=true
    ```

### XPath Syntax

The syntax provides a way to select nodes such as elements and attributes in an XML document.

Let's use the following payload:

```xml
<order>
    <customer id="1100">
        <name>John Doe</name>
        <email>john@example.com</email>
    </customer>
    <items>
        <item id="101">
            <name>Smartphone</name>
            <price>300</price>
        </item>
        <item id="102">
            <name>Laptop</name>
            <price>1200</price>
        </item>
    </items>
    <total>1500</total>
</order>
```

> The following table summarizes sample Xpath expressions and their outputs:

Reviewer: ```suggestion
The following table summarizes sample XPath expressions and their outputs:
```

<td><pre><code>1500</code></pre></td>
</tr>
<tr class="odd">
<td>Attributes</td>
<td>@: XPath can also be used to select attributes of elements.</td>
<td><pre><code>expression="//ns:customer/@id" xmlns:ns="http://ws.apache.org/ns/synapse"</code></pre></td>
<td><pre><code>1100</code></pre></td>
</tr>
<tr class="even">
<td>Predicates</td>
<td>Square Brackets ([]): Used to apply conditions to narrow down the node selection.</td>
<td><pre><code>expression="//ns:item[@id='102']/ns:name" xmlns:ns="http://ws.apache.org/ns/synapse"</code></pre></td>
<td><pre><code>Laptop</code></pre></td>
</tr>
<tr class="odd">
<td>Functions</td>
<td>XPath includes various built-in functions to work with nodes and data. Refer to <a href="https://www.w3schools.com/xml/xsl_functions.asp">more functions</a>.</td>
<td><pre><code>expression="sum(//ns:item/ns:price)" xmlns:ns="http://ws.apache.org/ns/synapse"</code></pre></td>
<td><pre><code>1500.0</code></pre></td>
</tr>
</tbody>
</table>

??? note "How to retrieve the XML node itself instead of text content"
    When the result of an XPath evaluation results in a single XML node, the

>     evaluator will return the text content of this node (equivalent of doing `/root/body/node/text()`). If you want to retrieve

Reviewer: ```suggestion
    evaluator will return the text content of this node (similar to executing `/root/body/node/text()`). If you want to retrieve
```

    in the following example.
    ``` xml
    <inSequence>
        <log level="custom">
            <property name="WHERE" value="before doing stuff"/>
        </log>
        <enrich>
            <source type="body" clone="true"/>
            <target type="property" property="ORIGINAL_PAYLOAD"/>
        </enrich>
        <property name="CHILD_PAYLOAD" expression="$body/child::node()" scope="default"/>
        <log level="custom">
            <property name="WHERE" value="after doing stuff"/>
            <property name="ORIGINAL_PAYLOAD" expression="get-property('ORIGINAL_PAYLOAD')"/>
            <property name="CHILD_PAYLOAD" expression="get-property('CHILD_PAYLOAD')"/>
        </log>
        <enrich>
            <source type="property" clone="true" property="ORIGINAL_PAYLOAD"/>
            <target type="body" action="sibling"/>
        </enrich>
        <log level="full"/>
    </inSequence>
    ```

### XPath extension functions

In addition to standard [XPath functions](https://www.w3schools.com/xml/xsl_functions.asp), the Micro Integrator supports the following custom functions for
working with XPath expressions:

#### base64Encode() function

> The `base64Encode` function returns the base64-encoded value of the

Reviewer: ```suggestion
The `base64Encode()` function returns the base64-encoded value of the
```

specified string.

Syntax:

-   `           base64Encode(string value)          `
-   `           base64Encode(string value, string charset)          `

Example:

`<property name="ENCODED_VALUE" expression="base64Encode('This is a test value')"/>`

#### base64Decode() function

> The `base64Decode` function returns the original value of the specified

Reviewer: ```suggestion
The `base64Decode()` function returns the original value of the specified
```

base64-encoded value.

The syntax of the function takes the following format.

-   `           base64Decode(string encodedValue)          `
-   `           base64Decode(string           encodedValue           , string charset)          `

Example:

`<property name="DECODED_VALUE" expression="base64Decode('VGhpcyBpcyBhIHRlc3QgdmFsdWU=')"/>`

#### get-property() function

The `get-property()` function allows you to retrieve various types of properties from the message context, such as transport headers, custom properties, system properties, or Axis2-specific properties. These properties can be used in the message flows, making the mediation process dynamic.

The syntax of the function takes the following format.

-   `          get-property(String propertyName)         `
-   `          get-property(String scope, String propertyName)         `

The function accepts [scope]({{base_path}}/reference/synapse-properties/scopes) as an optional parameter. It retrieves a
message property at the given scope. If you provide only the property name without the scope, the default `synapse` scope will be used.

#### url-encode() function

> The url-encode function returns the URL-encoded value of the specified

Reviewer: ```suggestion
The `url-encode()` function returns the URL-encoded value of the specified
```
